scalyr_agent/builtin_monitors/kubernetes_monitor.py

```diff
@@ -0,0 +1,1226 @@
+
+        #stop the DockerLoggers
+
+        for logger in self.raw_logs:
+            path = logger['log_config']['path']
+            if self.__log_watcher:
+                self.__log_watcher.remove_log_path( self.__module, path )
+            self._logger.log(scalyr_logging.DEBUG_LEVEL_1, "Stopping %s" % (path) )
+
+        self.raw_logs = []
+
+    def _process_namespaces( self, namespaces ):
+        """ Iterates over all namespaces on this node and then queries
+        every pod in every namespace.
+
+        @param namespaces: The JSON object returned as a response from querying
+            all namespaces from the k8s API
+
+        @return A dict of PodInfo objects, keyed on docker container ID
+        """
+        # get the list of namespaces
+        items = namespaces.get( 'items', [] )
+        containers = {}
+
+        # iterate over all namespaces, querying and processing all pods
+        # in the namespace
+        for namespace in items:
+            metadata = namespace.get( 'metadata', {} )
+            if 'name' in metadata:
+                pods = self.__k8s.query_pods( metadata['name'] )
+                self._process_pods( pods, containers )
+
+        return containers
+
+    def _process_pods( self, pods, containers ):
+        """ Iterates over every pod, storing PodInfo in the containers dict, hashed by
+            docker container id.
+
+            @param pods: The JSON object returned as a response from quering all pods
+                in a specific namespace
+
+            @param containers: A list of containers to add PodInfo to
+        """
+
+        # get all pods
+        items = pods.get( 'items', [] )
+
+        # iterate over all pods, getting PodInfo and storing it in the container
+        # dict, hashed by container ID
+        for pod in items:
+            info = self._process_pod( pod )
+
+            # The JSON object indicates that a Pod can have multiple
+            # containers, so make sure to key each container with this
+            # PodInfo
+            for container in info.containers:
+                containers[container] = info
+
+    def _process_pod( self, pod ):
+        """ Generate a PodInfo object from a JSON object
+        @param pod: The JSON object returned as a response to querying
+            a specific pod from the k8s API
+
+        @return A PodInfo object
+        """
+
+        result = {}
+
+        metadata = pod.get( 'metadata', {} )
+        spec = pod.get( 'spec', {} )
+        labels = metadata.get( 'labels', {} )
+
+        status = pod.get( 'status', {} )
+        containerStatuses = status.get( 'containerStatuses', [] )
+
+        containers = []
+
+        # build list of container ids used by this pod, making sure to
+        # strip docker:// from the front of each one
+        prefix = 'docker://'
+        for container in containerStatuses:
+            if 'containerID' not in container:
+                continue
```

Comment:
Is this an error?  If so, then let's emit a log line with a limit of one per 5 min